### ISSUE TITLE
BUG : PngExporter generates empty images

### ISSUE THREAD

**Issue by DanieleDef** [oxyplot/oxyplot#1498]
PngExporter produces an empty output If the user that started the application is not currently logged in.

### Steps to reproduce

1. Log into a remote PC
2. Start the application that periodically saves a png file using
```
var pngExporter = new PngExporter { Width = 2000, Height = 1000, Background = OxyColors.White };
            pngExporter.ExportToFile(pm, filepath);
```
3. logout from the remote pc
4. Check that the images stored when the remote connection is active are correctly saved and the one without an active remote connection are blank.

Platform:
.NET version: 4.5 (windows form)
Version oxyplot 2.0

### Expected behaviour

The export file should be correctly exported regardless of the remote connection state.

### Actual behaviour

If a remote connection is active or the user directly log in into the machine all works fine. If the user that started the application is not currently logged in all exported images are blank.

**Comment by Jonarw** [comment]
To be clear: Does this apply to `OxyPlot.Wpf.PngExporter` or `OxyPlot.ImageSharp.PngExporter`?
Did you try the respective other one?

**Comment by DanieleDef** [comment]
I'm using OxyPlot.Wpf.PngExporter.
It seems that I can't use the ImageSharp as I'm on the .NET 4.5. 
I've also tried the pdf exporter and it seems to work, but this is not a  suitable solution as I would need an image file.

```
var pngExporter2 = new OxyPlot.Wpf.PngExporter { Width = 2000, Height = 1000, Background = OxyColors.White };
            pngExporter2.ExportToFile(pm, @"C:\Temp\test3.jpg");
```

**Comment by Jonarw** [comment]
Since you are on WinForms, you could also give `OxyPlot.WindowsForms.PngExporter` a try (I only just realized this exists).

**Comment by DanieleDef** [comment]
I did try it with the same result.

**Comment by VisualMelon** [comment]
Is it possible to determine whether an exception is thrown during rendering (e.g. check `PlotModel.GetLastPlotException`)? Apparently there are some quirks in the GDI+ to do with remoting (e.g. [Clear](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.graphics.clear?redirectedfrom=MSDN&view=netframework-4.8#System_Drawing_Graphics_Clear_System_Drawing_Color_) can throw under some circumstances and presumably break the graphics object).

**Comment by DanieleDef** [comment]
The last plot exception is null before and after the export command.
I've tried with a transparent background but it doesn't help.

**Comment by ericjut1972** [comment]
FWIW, the CsvDemo in WpfExamples\Examples repro this issue for both xaml and png export, but svg and pdf are looking right.

![test2](https://user-images.githubusercontent.com/63374271/78835274-b8c90280-79a4-11ea-8ac7-2ac7b6ffc6b2.png)

**Comment by smetis** [comment]
i can see this behaviour is introduced into the code with changes for "Added high DPI scaling support for WPF"

**Comment by Jonarw** [comment]
Interesting, then I think it could possibly be already solved by #1504 (at least for WPF). I did not make the connection earlier, but I did notice that the DPI was not set to the `CanvasRenderContext` when using the `PngExporter` and changed that in that PR. Could you try again with the latest version?

**Comment by smetis** [comment]
i actually started out with latest code and saw this same issue, then back track on older commits to locate the commit that started to show this issue. so yes this issue should still occur in latest version.

**Comment by Jonarw** [comment]
I'm guessing you are using the `PlotView.SaveBitmap(...)` function?
As a workaround, could you try to use `new OxyPlot.Wpf.PngExporter().Export(...)`? I think that should work with the newest version of the code (after #1504).

**Comment by smetis** [comment]
now the latest code has the plot visible but the background again become black instead of white in previous code. 

![image](https://user-images.githubusercontent.com/6990404/79622880-0b559d80-80ce-11ea-931f-5c08d8be2358.png)

A side question, instead of making the background white as previous version, can we make png background transparent?

**Comment by Jonarw** [comment]
Weird... What is your `PlotModel.Background`? This should normally determine the background of the png. Can you share some code to reproduce this?

**Comment by smetis** [comment]
I actually test it out using the WPF example browser, just select any plots and then use Ctrl+C to copy the plot and paste it somewhere else. That should show this black background plot.

**Comment by VisualMelon** [comment]
Has the issue with generating empty images been resolved, such that we can close this issue?

<hr />

For the black background, if you think it is a bug, could you open another issue presenting the exact code, oxyplot version, and the resulting PNG? The WPF PNG exporter tests do not present any unexpected behaviour on my machine, so I'm not presently able to reproduce this.

**Comment by smetis** [comment]
Did you set the plotmodel background color in PngExporter test? PlotModel default background is {#00000000}, which makes the output bitmap bakcground black. To fix this issue, I have to explicitly set background to white when the background.IsVisible() return false in function BitmapSource ExportToBitmap(IPlotModel model).  

As stated in previous comment, this behavior only happen after recent merges. Not sure this is consider bug or just a change of behavior that PlotModel has to always set background color explicitly.

**Comment by ericjut1972** [comment]
I can't speak for the others, but I've tried to repro my issue (csvdemo rendering empty graphs with png and xaml outputs) with the latest bits and I'm not reproing it.

Thanks for fixing it!

**Comment by VisualMelon** [comment]
@smetis yes, I tried various `Background` settings. If I'm understanding correctly, the existing [`ExportToFile`](https://github.com/oxyplot/oxyplot/blob/develop/Source/OxyPlot.Wpf.Tests/PngExporterTests.cs#L39-L51) WPF PngExporter test is as you describe. At the release of 2.0, this would produce a white background, because this was the explicit defaulse in the WPF `PngExporter`; it has since changed to defaulting to transparent, as the exporter no longer has its own `Background` property.

Is there any possibility that your chosen viewer has a black background (e.g. the previewer in explorer), or renders base transparency as solid color? Could you upload the exact PNG it produces? If you have a piece of code that produces the issue, could you upload that as well (because I can't reproduce this) and check which versions you are using (i.e. are you using the 2.0 release package, or some later commit).

**Comment by DanieleDef** [comment]
I could not try it at the moment as the current arrangement uses Nuget to get the library. When is the next release planned?

**Comment by VisualMelon** [comment]
Release tend to be pretty rare. Can you use an AppVeyer CI pre-release build from [myget](https://www.myget.org/F/oxyplot/api/v3/index.json)? Version `2.1.0-unstable.1390` is the latest: something seems to have changed with the pre-release naming at some point, so VS seems to pick an earlier one by default.

**Comment by DanieleDef** [comment]
I have tested with the version :
- OxyPlot.Core.2.1.0-unstable.1390
- OxyPlot.WindowsForms.2.1.0-unstable.1390
- OxyPlot.Wpf.2.1.0-unstable.1390

I can confirm that the issue is still there for both export as .png and .jpg.

To make sure that the old libraries were not used anymore I deleted all old versions inside the package manager and manually compared the libraries in use with the one I'm expecting to use.

**Comment by Jonarw** [comment]
I think we have two separate issues here:

- The one experienced by ericjut1972 - the symptom here was that lines were not rendered when using the `CanvasRenderContext` for exporting without having it attached to a PlotView. This seems to be related to the High-DPI changes and should be fixed in the latest builds since #1504.

- The one experienced by DanieleDef - where the symptom is that completely empty images are being written when the Windows user running the program is not signed in. I think this ultimately comes down to the fact that drawing operations of WindowsForms and WPF UI frameworks don't seem to work properly when not signed in.

I am not too sure we can really fix this, this might be an inherent limitation of these UI frameworks.

You might want to give the new `OxyPlot.SkiaSharp.PngExporter` a try, this should work under .NET 4.5 and is independent of Winforms and WPF.

**Comment by DanieleDef** [comment]
I can confirm that the use of OxyPlot.SkiaSharp.PngExporter solves the issue. thank you.

**Comment by DanieleDef** [comment]
I've been able to export the image but the legend disappeared now. 
I'm using 
- OxyPlot.Core.2.1.0-unstable.1391
- OxyPlot.SkiaSharp.2.1.0-unstable.1391
It seems that most of the legend configuration are not visible anymore.

**Comment by VisualMelon** [comment]
I can't reproduce this with a simple example:

            var plot = new PlotModel() { Title = "Text" };
            plot.Series.Add(new FunctionSeries(x => Math.Sin(x), -10, 10, 0.1, "Sin"));
            plot.Series.Add(new FunctionSeries(x => Math.Cos(x), -10, 10, 0.1, "Cos"));
            plot.Legends.Add(new OxyPlot.Legends.Legend());
            OxyPlot.SkiaSharp.PngExporter.Export(plot, "test.png", 500, 500);

Could you provide an example, and elaborate on the platform (Framework vs. core, Win 7/10, etc.)? Is this issue still present with https://www.myget.org/feed/oxyplot/package/nuget/OxyPlot.Core/2.1.0-unstable.1405 ?

**Comment by DanieleDef** [comment]
I confirm that the example is working. The issue was caused by the missing instruction to include the legend that in the previous version was not necessary. Can I close the issue?

**Comment by VisualMelon** [comment]
@DanieleDef I'd be tempted to keep it open since the issue with Winforms and WPF isn't resolved, but I suppose someone else can reopen it in the future if they want.

**Comment by DanieleDef** [comment]
The workaround is fine for me but, as you said, the issue persists.
